### PR TITLE
cclient: added local PoW in attach_to_tangle. (WIP)

### PR DESCRIPTION
**Pull request by oopsmonk**
This PR implemented local PoW in iota_client_attach_to_tangle.

# Test Plan:
No.

### CODE REVIEW

File: cclient/iota_client_core_api.c [hunk 1]
```diff
@@ -432,33 +434,93 @@ retcode_t iota_client_attach_to_tangle(
...
-    goto done;
+  flex_trit_t prev_tx_hash[FLEX_TRIT_SIZE_243];
+  iota_transaction_t tx = {};
+  char_buffer_t* res_buff = NULL;
+  char_buffer_t* req_buff = NULL;
+  flex_trit_t* array_elt = NULL;
+  flex_trit_t* nonce = NULL;
+  flex_trit_t* raw_tx = NULL;
+
+  if (service == NULL) {
```
Comment (th0br0) [inline cclient/iota_client_core_api.c:441]: could we split this into two internal methods?

File: cclient/iota_client_core_api.c [hunk 2]
```diff
@@ -432,33 +434,93 @@ retcode_t iota_client_attach_to_tangle(
...
+    log_info(logger_id, "[%s:%d] local PoW\n", __func__, __LINE__);
+    memset(prev_tx_hash, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
+    HASH_ARRAY_FOREACH(req->trytes, array_elt) {
```
Comment (th0br0) [inline cclient/iota_client_core_api.c:448]: don't expect the trytes to be ordered?
(see my linked PR)

File: cclient/iota_client_core_api.c [hunk 3]
```diff
@@ -432,33 +434,93 @@ retcode_t iota_client_attach_to_tangle(
...
+        }
+        free(raw_tx);
+      } else {
+        log_info(logger_id,
+                 "[%s:%d] transaction serialize failed during PoW.\n", __func__,
+                 __LINE__);
+        return RC_CCLIENT_TX_DESERIALIZE_FAILED;
+      }
+
+      memcpy(prev_tx_hash, transaction_hash(&tx), FLEX_TRIT_SIZE_243);
```
Comment (th0br0) [inline cclient/iota_client_core_api.c]: no need to memcpy as the pointer won't go out of scope, just carry the ptr around

File: cclient/iota_client_core_api.c [hunk 4]
```diff
@@ -432,33 +434,93 @@ retcode_t iota_client_attach_to_tangle(
...
+      raw_tx = transaction_serialize(&tx);
```
Comment (th0br0) [inline cclient/iota_client_core_api.c]: Could we split the TX PoW and serialisation into distinct components?

e.g. cherrypick
https://github.com/th0br0/entangled/commit/7f21183126f9b642d5dd9a5dba47c57c5928ead1
https://github.com/th0br0/entangled/commit/a7dbf7ea5d515d12a828bfe32364d970c7ec0fdf

**Review by th0br0 — COMMENTED** [review]
Darn, now you beat me to just merging your impl!
rel https://github.com/th0br0/entangled/blob/mobile_expose_attachToTangle/common/helpers/pow.c#L84

**Review by th0br0 — CHANGES_REQUESTED** [review]
nt

File: cclient/iota_client_extended_api.c [hunk 5]
```diff
@@ -423,6 +423,7 @@ retcode_t iota_client_send_trytes(iota_client_service_t const* const serv,
                                   hash8019_array_p const trytes,
                                   uint32_t const depth, uint32_t const mwm,
                                   flex_trit_t const* const reference,
+                                  bool const local_PoW,
```
Comment (th0br0) [inline cclient/iota_client_extended_api.c]: all lowercase `local_pow` ?

File: cclient/iota_client_extended_api.h [hunk 6]
```diff
@@ -363,6 +363,7 @@ retcode_t iota_client_send_transfer(iota_client_service_t const* const serv,
  * @param {int} depth - Depth
  * @param {int} mwm - Min weight magnitude
  * @param {trit_array_p} reference - Optional reference hash
+ * @param {bool} local_PoW- apply local PoW
```
Comment (th0br0) [inline cclient/iota_client_extended_api.h]: there's a `-` too many here, also lowercase argument name

File: common/helpers/pow.c [hunk 7]
```diff
@@ -11,69 +11,126 @@
...
+    if (nonce == NULL) {
+      return RC_OOM;
+    }
+    transaction_set_nonce(tx, nonce);
+    free(nonce);
+
+    if (ctrunk != trunk) {
+      free(ctrunk);
+    }
+
```
Comment (th0br0) [inline common/helpers/pow.c:130]: we'll need to free & serialize the tx again as `iota_pow_flex` doesn't do inplace changes.

```
free(txflex);
txflex = transaction_serialize(tx);
```
then do the digest

we really need to revisit this typing fubar we have here soon (note to self)

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/helpers/pow.c [hunk 8]
```diff
@@ -11,69 +11,126 @@
...
+    // Find current tx
+    for (tx = (iota_transaction_t *)utarray_front(bundle);
+         tx != NULL && tx->essence.current_index != cur_idx;
+         tx = (iota_transaction_t *)utarray_next(bundle, tx))
+      ;
+
+    // Set trunk & branch
+    transaction_set_trunk(tx, ctrunk);
+    transaction_set_branch(tx, branch);
```
Comment (wusyong) [inline common/helpers/pow.c:108]: `transaction_set_attachment_timestamp` and upper/lower bound are missing.

Comment (oopsmonk) [reply common/helpers/pow.c:108]: thanks :)

File: common/helpers/pow.c (hunk 7 above)
Comment (oopsmonk) [reply common/helpers/pow.c:130]: txflex is freed at line 132. agree it needs to have a better approach.

File: cclient/iota_client_extended_api.c [hunk 9]
```diff
@@ -456,7 +457,11 @@ retcode_t iota_client_send_trytes(iota_client_service_t const* const serv,
   attach_req->mwm = mwm;
   attach_req->trytes = trytes;
   // attach to tangle
-  ret_code = iota_client_attach_to_tangle(serv, attach_req, attach_res);
+  if (local_pow) {
```
Comment (jserv) [inline cclient/iota_client_extended_api.c:460]: Why did't you rewrite as following?
```c
ret_code = iota_client_attach_to_tangle(!localpow ? serv : NULL, attach_req, attach_res);
```

File: common/helpers/pow.c [hunk 10]
```diff
@@ -11,69 +11,129 @@
...
+
+    if (ctrunk != trunk) {
+      free(ctrunk);
+    }
+
+    ctrunk = iota_flex_digest(txflex, NUM_TRITS_SERIALIZED_TRANSACTION);
+    free(txflex);
+  } while (cur_idx != 0);
+
+  if (ctrunk != trunk && ctrunk != NULL) {
```
Comment (jserv) [inline common/helpers/pow.c]: The condition `ctrunk != NULL` is not necessary because `free(NULL)` is no-op.

**Review by th0br0 — DISMISSED** [review]

**Review by th0br0 — APPROVED** [review]

**Pull request merged**